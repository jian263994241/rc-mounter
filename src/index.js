import React, {Component, isValidElement, cloneElement, createElement} from 'react'
import {render, unmountComponentAtNode, createPortal, unstable_renderSubtreeIntoContainer} from 'react-dom'
import PropTypes from 'prop-types'

export default class Mounter extends Component {

  static uiName = 'Mounter';

  static propTypes = {
    component: PropTypes.oneOfType([PropTypes.element , PropTypes.string])
  };

  static defaultProps = {
    component: 'div'
  };

  container = null;

  getContainer = ()=>{
    if(this.container) return this.container;
    let container = document.createElement('div');
    this.container = container;
    document.body.appendChild(container);
    return container;
  };

  getComponent = ()=>{
    const props = this.props;
    const {prefixCls, component, ...rest} = this.props;

    if(isValidElement(component)){
      return cloneElement(component, {...rest})
    }

    return createElement(component, rest);
  }

  removeContainer = () => {
    const container = this.container;
    if(!container) return ;
    if(!createPortal){ unmountComponentAtNode(container); }
    container.parentNode.removeChild(container);
  };

  renderComponent(){
    if(!createPortal){
      unstable_renderSubtreeIntoContainer(this, this.getComponent(), this.getContainer());
    }
  }

  componentDidMount(){
    this.renderComponent();
  }

  componentDidUpdate(){
    this.renderComponent();
  }

  componentWillUnmount(){
    this.removeContainer();
  }

  render() {

    if(createPortal){
      return createPortal(this.getComponent(), this.getContainer());
    }

    return null;
  }

}
