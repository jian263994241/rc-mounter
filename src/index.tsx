import * as React from 'react';
import * as ReactDOM from 'react-dom';
import PropTypes, { InferProps } from 'prop-types';

type Container = React.Component | HTMLElement;

function getContainer(container?: Container | (() => Container)): HTMLElement {
  // #StrictMode ready
  return ReactDOM.findDOMNode(
    typeof container === 'function' ? container() : container
  ) as HTMLElement;
}

const useEnhancedEffect =
  typeof window !== 'undefined' ? React.useLayoutEffect : React.useEffect;

const PortalPropTypes = {
  /**
   * The children to render into the `container`.
   */
  children: PropTypes.element,
  /**
   * A node, component instance, or function that returns either.
   * The `container` will have the portal children appended to it.
   * By default, it uses the body of the top-level document object,
   * so it's simply `document.body` most of the time.
   */
  container: PropTypes.oneOfType([
    PropTypes.func,
    PropTypes.instanceOf(React.Component),
    PropTypes.instanceOf(Element)
  ]),
  /**
   * Disable the portal behavior.
   * The children stay within it's parent DOM hierarchy.
   */
  disabled: PropTypes.bool,
  /**
   * Callback fired once the children has been mounted into the `container`.
   *
   */
  onRendered: PropTypes.func
};

const Portal: React.FC<InferProps<typeof PortalPropTypes>> = React.forwardRef(
  (props, ref) => {
    const { children, disabled, onRendered, container = document.body } = props;

    const [mountNode, setMountNode] = React.useState<HTMLElement | null>(null);

    useEnhancedEffect(() => {
      if (!disabled) {
        setMountNode(getContainer(container as Container));
      }
    }, [container, disabled]);

    useEnhancedEffect(() => {
      if (onRendered && (mountNode || disabled)) {
        onRendered();
      }
    }, [onRendered, mountNode, disabled]);

    if (disabled) {
      return children as React.ReactElement;
    }

    if (mountNode) {
      return ReactDOM.createPortal(children, mountNode);
    }

    return null;
  }
);

Portal.propTypes = PortalPropTypes;

Portal.displayName = 'Portal';

export default Portal;
export { Portal };
