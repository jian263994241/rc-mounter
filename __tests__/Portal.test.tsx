import '@testing-library/jest-dom';
import React from 'react';
import { render } from '@testing-library/react';
import { Portal } from '../src';

describe('Portal', () => {
  it('default props', () => {
    function A() {
      return (
        <Portal>
          <div>123123</div>
        </Portal>
      );
    }

    const { debug } = render(<A />);

    debug();
  });

  it('disabled props', () => {
    function A() {
      return (
        <Portal disabled>
          <div>123123</div>
        </Portal>
      );
    }

    const { debug } = render(<A />);

    debug();
  });
});
