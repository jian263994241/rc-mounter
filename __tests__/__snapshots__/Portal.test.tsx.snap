// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Portal should work in a simple case 1`] = `<div />`;
